### DISCUSSION TITLE
undefined reference to `openvdb::v7_2::initialize()'

### DISCUSSION THREAD

**Discussion by Micrufun** [AcademySoftwareFoundation/openvdb#1312] [Q&A] [answered]
On Linux, I'm building with OpenVDB by:

https://www.openvdb.org/documentation/doxygen/build.html#buildUsingOpenVDB

Like:

```cmake
    list(APPEND CMAKE_MODULE_PATH "/usr/local/lib64/cmake/OpenVDB/")
    find_package(OpenVDB REQUIRED)

    target_link_libraries(${PROJECT_NAME} PUBLIC
        OpenVDB::openvdb
    )
```
I've built OpenVDB by checking out the tag `v7.2.3`. I've also defined the macros by CMake:

```cmake
	add_compile_definitions(OPENVDB_7_ABI_COMPATIBLE)
```

Also by C++:

```c++
// https://github.com/AcademySoftwareFoundation/openvdb/issues/139
#define OPENVDB_7_ABI_COMPATIBLE
#include <openvdb/openvdb.h>
```
# Error

But I'm receiving this linker error when building my project:

> undefined reference to `openvdb::v7_2::initialize()'

A this line of C++ code:

```c++

    openvdb::FloatGrid::Ptr mesh_to_grid(
        const Contour &mesh, const openvdb::math::Transform &tr = {}, float exteriorBandWidth = 3.0f, float interiorBandWidth = 3.0f, int flags = 0)
    {
        openvdb::initialize(); // ***** At this line the linker error is thrown.
        return openvdb::tools::meshToVolume<openvdb::FloatGrid>(
            ContourDataAdapter{mesh}, tr, exteriorBandWidth, interiorBandWidth,
            flags);
    }
```
# Version 8

I tried checking out `v8.1.0` tag when building OpenVDB. But eventually a similar error is thrown:

> undefined reference to `openvdb::v8_1::initialize()'

I ran out of options to debug this error. Can anybody help?

# Note

Previously, I was linking to OpenVDB on a different machine just fine. But, installing dependencies and building on a new Linux machine throws the above errors.

# Update

Investigating CMake log indicates that linker is *not* able to link to any library. Including `tbb`, `openvdb`, ...

It may be a general problem. It's not specifically related to OpenVDB.

A similar problem is described here:

https://stackoverflow.com/q/11112318

> Linker cannot find symbols, but libraries are read and symbols exist

Root cause might be:

https://stackoverflow.com/questions/11112318/linker-cannot-find-symbols-but-libraries-are-read-and-symbols-exist#comment75498543_11112318

> Some recent distributions apparently default to using the --as-needed linker flag, which enforces that the program's object files come before the dynamic libraries. If that flag is passed, the linker will not link to libraries that are not actually needed by the executable (and it detects this from left to right).

**Comment by Megidd** [comment, marked as answer]
# Fix 🔧 

The cause of the linker errors was this statement:

```cmake
if(LINUX)
```

The fix was to replace it with this:

```cmake
if(UNIX AND NOT APPLE)
```

This commit fixes the problem:

[![Commit][1]][1]

Reference: https://stackoverflow.com/a/40152725/3405291

Strangely, CMake wasn't complaining about anything and just throwing random linker errors.

  [1]: https://i.stack.imgur.com/Bjxm2.png

**Reply by Idclip** [reply]
@Megidd was this in our CMake? I can't see this code anywhere

**Reply by Megidd** [reply]
@Idclip No worries, it was **_not_** in OpenVDB code :)

**Reply by astitvaveergarg** [reply]
wait , so where is it ?

**Reply by Megidd** [reply]
@astitvaveergarg It was inside my own code which was using OpenVDB as a dependency.
